timemachine/fe/rest/bond.py

```diff
@@ -0,0 +1,100 @@
+from abc import ABC, abstractmethod
```

Comment:
This module might eventually be generally applicable, but is limited to the `rest` submodule for now to avoid too large a diff. The idea is to introduce dataclasses representing canonical ixns, which can then be used as argument types to ensure on the type level that indices are canonicalized where expected. Similar vectorized version could be introduced to handle contexts where we have 2d ndarrays of indices.